### ISSUE TITLE
Token Verification Can Generate Incorrect Message 

### ISSUE THREAD

**Issue by ncvitak** [vapor/jwt#30]
Currently when verifying a token the message is rebuilt from the header and payload nodes via createMessage(). This only produces the correct message when the original JSON is ordered and minified in the exact same way, otherwise a malformed message is produced and verification will fail.

**Example** (https://jwt.io)

Source Header:
```
{ 
  "alg": "HS256",
  "typ": "JWT"
}
```
Source Payload:
```
{
  "sub": "1234567890",
  "name": "John Doe",
  "admin": true
}
```

Source Header Minified:
```
{"alg":"HS256","typ":"JWT"}
```

Source Payload Minified:
```
{"sub":"1234567890","name":"John Doe","admin":true}
```

Source Message:
``eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWV9``

VaporJWT Reconstructed Header:
```
{"alg":"HS256","typ":"JWT"}
```

VaporJWT Reconstructed Payload: ***(different than source)***
```
{"admin":true,"name":"John Doe","sub":"1234567890"}
```

VaporJWT Reconstructed Message: ***(different than source)***
``eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhZG1pbiI6dHJ1ZSwibmFtZSI6IkpvaG4gRG9lIiwic3ViIjoiMTIzNDU2Nzg5MCJ9``

**Comment by tanner0101** [comment]
Does the JWT spec require the payload key order be maintained?

**Comment by ncvitak** [comment]
If you modify the key order, have you not modified the JSON, and thus the token you are attempting to verify?

My point is that this fails the basic example at https://jwt.io, and also fails when verifying [Firebase](https://firebase.google.com) tokens for the same reason.

One way to fix this would be to privately store the original message bytes from the decoded token in the JWT struct. This would only be set if the JWT was initialized from a Token string, otherwise it would be nil. Upon verification it'd use the stored original message if it exists, otherwise it can use createMessage() since the token must have been generated by Vapor anyways.

**Comment by vzsg** [comment]
Let me help clarify the issue.

The following test case explodes with signature verification error:

```swift
    func testHS256VerificationOfWellKnownToken() throws {
        let jwt = try JWT(token: "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJPbmxpbmUgSldUIEJ1aWxkZXIiLCJpYXQiOjE0ODkwMDE0MzIsImV4cCI6MTQ4OTAwMTQzMSwiYXVkIjoiIiwic3ViIjoiMTIzNDU2Nzg5MCIsIm5hbWUiOiJKb2huIERvZSIsImFkbWluIjoidHJ1ZSJ9.wbpKZ72B1heiE_iRnv82FjX7AfCLQ5lVsYR60t8bomw")

        try jwt.verifySignature(using: HS256(key: "foobar".bytes))
    }
```

And the **reason** behind the failure (as correctly deduced by @ncvitak) is that the library is not validating the token string, but a fake, recomputed version inside the JWT class.

**Comment by tanner0101** [comment]
Aha, I see. The signature is created from the bytes `{"a":"a", "b":"b"}` to technically `{"b":"b", "a":"a"}` is not valid against the signature. 

Possibly a fix could be validating the tokens within the `try JWT(token: ` init method. That way we still have access to the original message.

We could change it to `try JWT(validatingToken: xxx)` to be explicit.

**Comment by ncvitak** [comment]
Better to validate afterwards because some tokens, such as those from Firebase have a "kid" in the header that you use select the correct key for validation

**Comment by vzsg** [comment]
^ this. In such cases, the Signer cannot be created before parsing the JWT once. It would be better to store the JWT string directly (as `rawToken` for example), and perform signature tests and compute other properties from this.

**Comment by tanner0101** [comment]
Fix is ready to go here: https://github.com/vapor/jwt/pull/32

**Comment by vzsg** [comment]
The issue was resolved by that fix, released as 0.11.1 and 0.8.1, so I'm closing the issue.
